### PR TITLE
Improvements for failover of Distributed queries

### LINKED ISSUE #5317: Increase the speed of getting replica back to healthy state
Case
1) some replica go temporary offline
2) Distributed table connection pool error_count for that replica get high
3) replica gets back online again, but Distributed doesn't use it for a long time, as the speed of error_count decreasing is very slow currently (halved every 10 minutes).

We need to make some better heuristic for that. 

We can just decrease the value for
DBMS_CONNECTION_POOL_WITH_FAILOVER_DEFAULT_DECREASE_ERROR_PERIOD
but in that case flapping replica errors will be 'forgotten' too fast ( see https://github.com/yandex/ClickHouse/commit/192bbc02859ba140cb46878750318dbbc13f5837 ).

Maybe we should introduce some upper limit for `error_count` and never increase it above that number to give a chance to return back alive fast enough?

### PR DESCRIPTION
**Pull request by Enmk**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Improvement

Short description (up to few sentences):
Improvements for failover of Distributed queries. Shorten recovery time, also it is now configurable and can be seen in system.clusters.
...

Detailed description (optional):
* Added a limit on how many errors can replica accumulate
* Decreased default error halving time to 60 seconds
* Made both configurable via settings: `replica_error_max_count` and `replica_error_decrease_period`
* Showing errors count and estimated recovery time for each replica in `system.clusters`

## Example
```
SELECT *
FROM system.clusters

┌─cluster──────┬─shard_num─┬─shard_weight─┬─replica_num─┬─host_name─┬─host_address─┬─port─┬─is_local─┬─user────┬─default_database─┬─errors_count─┬─estimated_recovery_time─┐
│ test_cluster │         1 │            1 │           1 │ localhost │ 127.0.0.1    │ 9000 │        1 │ default │                  │            6 │                     166 │
│ test_cluster │         1 │            1 │           2 │ localhost │ 127.0.0.1    │ 9001 │        1 │ default │                  │            0 │                       0 │
│ test_cluster │         1 │            1 │           3 │ localhost │ 127.0.0.1    │ 9002 │        1 │ default │                  │            6 │                     166 │
└──────────────┴───────────┴──────────────┴─────────────┴───────────┴──────────────┴──────┴──────────┴─────────┴──────────────────┴──────────────┴─────────────────────────┘

```
* `errors_count` is number of times this host tried to reach replica but failed
* `estimated_recovery_time` how many seconds are left until replica error count is zeroed and replica is considered to be back to normal

Please note that `errors_count` is updated once per query to the cluster, but `estimated_recovery_time` is recalculated on-demand. So there could be a case of  non-zero `errors_count` and zero `estimated_recovery_time`, that means that next query will zero errors count and will try to use replica as if it had no errors.
...

Fixes #5317

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Let's merge with master, because we have fixes for integration and performance tests.

File: dbms/src/Core/Settings.h [hunk 1]
```diff
@@ -341,6 +341,8 @@ struct Settings : public SettingsCollection<Settings>
     \
     M(SettingUInt64, max_partitions_per_insert_block, 100, "Limit maximum number of partitions in single INSERTed block. Zero means unlimited. Throw exception if the block contains too many partitions. This setting is a safety threshold, because using large number of partitions is a common misconception.") \
     M(SettingBool, check_query_single_value_result, true, "Return check query result as single 1/0 value") \
+    M(SettingUInt64, replica_error_decrease_period, DBMS_CONNECTION_POOL_WITH_FAILOVER_DEFAULT_DECREASE_ERROR_PERIOD, "Time period reduces replica error counter by 2 times.") \
+    M(SettingUInt64, replica_error_max_count, DBMS_CONNECTION_POOL_WITH_FAILOVER_MAX_ERROR_COUNT, "Max number of errors per replica, prevents piling up increadible amount of errors if replica was offline for some time and allows it to be reconsidered in a shorter amount of time.") \
```
Comment (akuzm) [inline dbms/src/Core/Settings.h]: Looks like these settings are not used.

File: dbms/src/Core/Defines.h [hunk 2]
```diff
@@ -41,7 +41,9 @@
 #define DBMS_CONNECTION_POOL_WITH_FAILOVER_DEFAULT_MAX_TRIES 3
 /// each period reduces the error counter by 2 times
 /// too short a period can cause errors to disappear immediately after creation.
-#define DBMS_CONNECTION_POOL_WITH_FAILOVER_DEFAULT_DECREASE_ERROR_PERIOD (2 * DBMS_DEFAULT_SEND_TIMEOUT_SEC)
+#define DBMS_CONNECTION_POOL_WITH_FAILOVER_DEFAULT_DECREASE_ERROR_PERIOD 60
+/// replica error max cap
+#define DBMS_CONNECTION_POOL_WITH_FAILOVER_MAX_ERROR_COUNT 1000
```
Comment (akuzm) [inline dbms/src/Core/Defines.h]: No point in defines that are only used once.

Comment (Enmk) [reply dbms/src/Core/Defines.h]: I see your point, but well, it is actually used twice:
* ./dbms/src/Core/Settings.h:345
* ./dbms/src/Client/ConnectionPoolWithFailover.h:41

And also, IMO, that helps documenting what the value is and why, rather that having just some magic number.

File: dbms/src/Core/Settings.h (hunk 1 above)
Comment (Enmk) [reply dbms/src/Core/Settings.h]: Ouch, forgot to commit the crucial thing, thank you for noticing.

Comment (Enmk) [reply dbms/src/Core/Settings.h]: fixed

File: dbms/src/Client/ConnectionPoolWithFailover.h [hunk 3]
```diff
@@ -34,7 +37,8 @@ class ConnectionPoolWithFailover : public IConnectionPool, private PoolWithFailo
     ConnectionPoolWithFailover(
             ConnectionPoolPtrs nested_pools_,
             LoadBalancing load_balancing,
-            time_t decrease_error_period_ = DBMS_CONNECTION_POOL_WITH_FAILOVER_DEFAULT_DECREASE_ERROR_PERIOD);
+            time_t decrease_error_period_ = DBMS_CONNECTION_POOL_WITH_FAILOVER_DEFAULT_DECREASE_ERROR_PERIOD,
+            size_t max_error_cap = DBMS_CONNECTION_POOL_WITH_FAILOVER_MAX_ERROR_COUNT);
```
Comment (akuzm) [inline dbms/src/Client/ConnectionPoolWithFailover.h]: Does this mean that some callers use the pool with compile-time settings, and some -- with user-specified settings? What is the reason for that?

Comment (Enmk) [reply dbms/src/Client/ConnectionPoolWithFailover.h]: `ClickHouseDictionarySource` uses defaults, the main reason is that these settings are for controlling distributed tables / replicas even the setting name begins with `replica_*`. I am not sure if error cap/recovery rate of remote dictionaries should be configurable, but it is definitely a bad idea to use same settings for both dictionaries and cluster replicas.

Comment (akuzm) [reply dbms/src/Client/ConnectionPoolWithFailover.h]: OK, I see. I'd make the parameters explicit so that the caller doesn't forget to set the right values, but that's up to you.

There is another use in StorageDistributedDirectoryMonitor::createPool that has default timeouts. Not sure what it's used for, is this OK?

File: dbms/src/Core/Defines.h (hunk 2 above)
Comment (akuzm) [reply dbms/src/Core/Defines.h]: OK, somehow I thought that the second use is part of the forgotten commit.

File: dbms/src/Client/ConnectionPoolWithFailover.h (hunk 3 above)
Comment (Enmk) [reply dbms/src/Client/ConnectionPoolWithFailover.h]: Fixed that, also updated `replica_error_decrease_period` to be `SettingSeconds` instead of `SettingUInt64`.

File: dbms/src/Core/Settings.h [hunk 4]
```diff
@@ -345,6 +345,9 @@ struct Settings : public SettingsCollection<Settings>
     M(SettingBool, check_query_single_value_result, true, "Return check query result as single 1/0 value") \
     M(SettingBool, allow_drop_detached, false, "Allow ALTER TABLE ... DROP DETACHED PART[ITION] ... queries") \
     \
+    M(SettingSeconds, replica_error_decrease_period, DBMS_CONNECTION_POOL_WITH_FAILOVER_DEFAULT_DECREASE_ERROR_PERIOD, "Time period reduces replica error counter by 2 times.") \
+    M(SettingUInt64, replica_error_max_count, DBMS_CONNECTION_POOL_WITH_FAILOVER_MAX_ERROR_COUNT, "Max number of errors per replica, prevents piling up increadible amount of errors if replica was offline for some time and allows it to be reconsidered in a shorter amount of time.") \
```
Comment (akuzm) [inline dbms/src/Core/Settings.h]: We probably should prefix the names with something like 'distributed' to emphasize that these parameters control the Distributed table engine, and to avoid confusion with ReplicatedMergeTree replicas, Dictionary replicas and maybe some other replicas we might have.

About the names themselves, I'd name the first one 'distributed_replica_error_half_life'. This name makes it immediately clear that the error count decays exponentially, and the parameter controls the time after which the error count is reduced twice.
For the second one, I'd use 'distributed_replica_error_cap', to convey that the error count is capped at this value. 'max_count' is less precise, it may mean for example that the replica is disconnected after this number is reached, or something else.

**Comment by akuzm** [comment]
The docs should be updated to describe the new settings and the new fields in system.clusters.

File: dbms/src/Core/Settings.h (hunk 4 above)
Comment (Enmk) [reply dbms/src/Core/Settings.h]: Agree, fixed

File: docs/en/operations/system_tables.md [hunk 5]
```diff
@@ -97,19 +107,19 @@ Contains information about external dictionaries.
...
-- `attribute.types Array(String)` — Corresponding array of attribute types that are provided by the dictionary.
-- `has_hierarchy UInt8` — Whether the dictionary is hierarchical.
-- `bytes_allocated UInt64` — The amount of RAM the dictionary uses.
-- `hit_rate Float64` — For cache dictionaries, the percentage of uses for which the value was in the cache.
-- `element_count UInt64` — The number of items stored in the dictionary.
-- `load_factor Float64` — The percentage filled in the dictionary (for a hashed dictionary, the percentage filled in the hash table).
-- `creation_time DateTime` — The time when the dictionary was created or last successfully reloaded.
-- `last_exception String` — Text of the error that occurs when creating or reloading the dictionary if the dictionary couldn't be created.
-- `source String` — Text describing the data source for the dictionary.
+- `name` (String) — Dictionary name.
```
Comment (akuzm) [inline docs/en/operations/system_tables.md]: Let's merge the unrelated documentation change as another PR.

**Comment by Enmk** [comment]
Docs updated, settings renamed.

File: docs/en/operations/system_tables.md (hunk 5 above)
Comment (Enmk) [reply docs/en/operations/system_tables.md]: Ok, reverted that back, leaving only relevant documentation changes.

**Review by akuzm — APPROVED** [review]

**Review by alexey-milovidov — APPROVED** [review]
